qiskit_experiments/library/tomography/basis/local_basis.py

```diff
@@ -0,0 +1,414 @@
+
+        # Format instructions so compatible types can be converted to
+        # Instruction instances.
+        if instructions is not None:
+            self._instructions = _format_instructions(instructions)
+            self._size = len(instructions)
+            if default_states is None:
+                default_states = self._instructions
+                self._custom_defaults = False
+
+        # Construct default states
+        if default_states is not None:
+            self._default_states = tuple(DensityMatrix(i).data for i in default_states)
+            self._default_dim = self._default_states[0].shape[0]
+            if self._size is None:
+                self._size = len(self._default_states)
+            elif len(self._default_states) != self._size:
+                raise QiskitError(
+                    "Number of instructions and number of default states must be equal."
+                )
+
+        # Construct states of specific qubits if provided
+        qubit_states = qubit_states or {}
+        for qubit, states in qubit_states.items():
+            if self._size is None:
+                self._size = len(states)
+            elif len(states) != self._size:
+                raise QiskitError("Number of instructions and number of states must be equal.")
+
+            qstates = tuple(DensityMatrix(i).data for i in states)
+            self._qubit_states[qubit] = qstates
+            self._qubit_dim[qubit] = qstates[0].shape[0]
+            self._qubits.add(qubit)
+
+        # Pseudo hash value to make basis hashable for LRU cached functions
+        self._hash = hash(
+            (
+                type(self),
+                self._name,
+                self._size,
+                self._default_dim,
+                self._custom_defaults,
+                tuple(self._qubits),
+                tuple(self._qubit_dim.values()),
+                (type(i) for i in self._instructions),
+            )
+        )
+
+    def __repr__(self):
+        return f"<{type(self).__name__}: {self.name}>"
+
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, value):
+        return (
+            super().__eq__(value)
+            and self._size == getattr(value, "_size", None)
+            and self._default_dim == getattr(value, "_default_dim", None)
+            and self._custom_defaults == getattr(value, "_custom_defaults", None)
+            and self._qubits == getattr(value, "_qubits", None)
+            and self._qubit_dim == getattr(value, "_qubit_dim", None)
+            and self._instructions == getattr(value, "_instructions", None)
+        )
+
+    def index_shape(self, qubits: Sequence[int]) -> Tuple[int, ...]:
+        return len(qubits) * (self._size,)
+
+    def matrix_shape(self, qubits: Sequence[int]) -> Tuple[int, ...]:
+        return tuple(self._qubit_dim.get(i, self._default_dim) for i in qubits)
+
+    def circuit(self, index: Sequence[int], qubits: Sequence[int]) -> QuantumCircuit:
```

Comment:
To me basis having circuit looks bit weird, i.e. mixture of data and program. Can we delegate circuit synthesis to tobmogprahy circuit method? Then basis will become a simple container-like object that just stores operators. I don't think this hurts readability of the tomography experiment method.

Reply:
> you would need to generate different POVM matrices based on measurement error

To me this still sounds like the basis is still container of model, ie POVM elements. However,

> The other purpose is for analysis to define the matrices those preparation/measurement circuits correspond to given a model for the system

This makes me think current design is reasonable. This indicates operators (i.e. models) and implementation (i.e. circuits) are not always identical. So sometimes you want to assign different models to interpret the program. That makes sense. Can you add this to the base class documentation for developer? This explanation makes clearer why basis can have partial circuits though experiment creates circuits.